### PR TITLE
[GVN] Permit load PRE to happen in more cases

### PR DESCRIPTION
**Pull request by david-arm**
There is code in PerformLoadPRE that bails out if the number of
predecessors where we have to insert a load is greater than one:

```
  // If we need to insert new load in multiple predecessors, reject it.
  // FIXME: If we could restructure the CFG, we could make a common pred with
  // all the preds that don't have an available Load and insert a new load into
  // that one block.
  if (NumInsertPreds > 1)
      return false;
```

However, this is quite restrictive because NumInsertPreds is the sum

```
  unsigned NumInsertPreds = PredLoads.size() + CriticalEdgePredSplit.size();
```

where PredLoads and CriticalEdgePredSplit are the lists of available and
unavailable loads, respectively. Therefore, with just one unavailable
load we refuse to perform load PRE if there are other predecessors with
available loads. I've changed this check above to be less conservative:

```
  if (CriticalEdgePredSplit.size() > 1)
      return false;
```

As a result we can perform load PRE in many more cases. On the
AArch64 neoverse-v1 machine we see 1% gains in both SPEC2017 perlbench
and omnetpp benchmarks. On neoverse-n1 we see 1% and 3.5% gains in
perlbench and omnetpp, respectively. I only observed one regression with
exchange2, but this is related to function specialisation and will be
fixed separately.

### CODE REVIEW

**Review by nikic — CHANGES_REQUESTED** [review]
> where PredLoads and CriticalEdgePredSplit are the lists of available and
unavailable loads, respectively.

This is incorrect. Both are lists of unavailable loads. The first one for unavailable loads that don't require critical edge splitting, and the second for ones that do.

> Therefore, with just one unavailable load we refuse to perform load PRE if there are other predecessors with available loads.

And this conclusion is thus also incorrect.

What the current check does it to make sure that the number of loads does not increase. We end up removing one load, and adding one load in an unavailable block.

Your new check says that instead we can insert an arbitrary number of loads in unavailable predecessors, as long as there is at least one available load (and we don't have to split more than one critical edge). This is not a reasonable heuristic.

**Comment by david-arm** [comment]
> Your new check says that instead we can insert an arbitrary number of loads in unavailable predecessors, as long as there is at least one available load (and we don't have to split more than one critical edge). This is not a reasonable heuristic.

Fair enough! But I would like to fix the problem highlighted in test7a, which is definitely blocked by this check. Given the potential performance gains, I'd still like to pursue this and I'm happy to do more detailed analysis. What would you consider a reasonable heuristic?

File: llvm/test/Transforms/GVN/PRE/load-pre-licm.ll [hunk 1]
```diff
@@ -282,3 +284,97 @@ header:
   call void @hold(i32 %v1)
   br label %header
 }
+
+
+; In block Z one load is unavailable (via entry block), whereas via other 2
+; predecessors the loads are available.
```
Comment (nikic) [inline llvm/test/Transforms/GVN/PRE/load-pre-licm.ll:290]: As far as GVN knows, the load is not available in block B. If it were, the transform would already be performed.

**Comment by david-arm** [comment]
@nikic So perhaps I've misunderstood what this comment in the code above means:

```
  // Check to see how many predecessors have the loaded value fully
  // available.
  MapVector<BasicBlock *, Value *> PredLoads;
```

I took that to mean that `PredLoads` contains the loads from all predecessors that are fully available.

**Comment by nikic** [comment]
> > Your new check says that instead we can insert an arbitrary number of loads in unavailable predecessors, as long as there is at least one available load (and we don't have to split more than one critical edge). This is not a reasonable heuristic.
> 
> Fair enough! But I would like to fix the problem highlighted in test7a, which is definitely blocked by this check. Given the potential performance gains, I'd still like to pursue this and I'm happy to do more detailed analysis. What would you consider a reasonable heuristic?

You should start by trying to understand why GVN does not recognize the available load in that test case. It probably has something to do with the the fact that one block contains two available loads for different phi translated pointers.

**Comment by david-arm** [comment]
> You should start by trying to understand why GVN does not recognize the available load in that test case. It probably has something to do with the the fact that one block contains two available loads for different phi translated pointers.

OK thanks. I think I see now that `PredLoads` is different to what I thought. If I understand correctly, `IsValueFullyAvailableInBlock` should return true for my test, which would then prevent `PredLoads` from being populated.

**Comment by david-arm** [comment]
I've created https://github.com/llvm/llvm-project/pull/79324 that does the analysis to add more loads to the available list.

**Pull request closed without merging**